## COSC 1210
### print(), input(), value storage and chaos

Objectives
- use the chaos function to illustrate the importance of value storage

In [1]:

def main():
    a = 1/3
    b = a + a + a
    print(b)
main()

1.0


##### Explain why the function above provides an answer that is different than the algebraic solution

The variable is rounded to 17 decimal points so the answer is not quite accurate.

In [2]:
def main():
    a = 1/7
    b = a * 7
    c = a + a + a + a + a + a + a
    print("b is", b)
    print("c is", c)
main()

b is 1.0
c is 0.9999999999999998


##### Explain why b and c provide different answers

B rounds the variable less times than c does. This causes more inaccuracies to build up causing a difference in value.

##### How large is the difference between c and b?

0.0000000000000002



### rounding floats
Mathematically, what will be the value of b?

It would be 1

When calculated by python, what is the value of b?

Since it rounded automatically it would be 1 and unrounded would be about 0.9999999999998

In [4]:
def main():
    print("This program illustrates a chaotic function")
    iterations = 100
    x1 = 0.15
    for i in range(iterations):
        x1 = 3.9 * x1 * (1 - x1)
    print("For", iterations, "iterations, x1 is:", x1)
main()

This program illustrates a chaotic function
For 100 iterations, x1 is: 0.5126527786207501


##### If you calculate x1 for 10 iterations vs 100 iterations, what answers do you get?

for 10 iterations: 0.9735732406265619

for 100 iterations: 0.5126527786207501

##### What are these different?
They are different because they are rounded a different number of times. Since 100 iterations uses the variable rounding a lot, it will become less accurate.


In [8]:
def main():
    print("This program illustrates a chaotic function")
    iterations = 10
    x1 = 0.15
    x2 = 0.15
    for i in range(iterations):
        x1 = 3.9 * x1 * (1 - x1)
        x2 = 3.9 * (x2 - x2 * x2)
    print("For", iterations, "iternations, x1 is now", x1, "and x2 is now", x2)
main()

This program illustrates a chaotic function
For 10 iternations, x1 is now 0.9735732406265619 and x2 is now 0.9735732406265659


##### If you calculate x1 and x2 for 10 iterations vs 100 iterations, what answers do you get?

x1 10 iterations: 0.9735732406265619 

x2 10 iterations: 0.9735732406265659



x1 100 iterations: 0.5126527786207501 

x2 100 iterations: 0.16411537941360743

##### Why are these different?
These are different because x2 uses the variable more times than x1. This causes more inaccuracies compared to x1 when having the same amount of iterations.

In [2]:
### Rewrite the chaos function so that the user can enter x (iterations?)
def main():
    print("This program illustrates a chaotic function")
    userinput = input("Enter number of iterations: ")
    iterations = eval(userinput)
    x1 = 0.15
    x2 = 0.15
    for i in range(iterations):
        x1 = 3.9 * x1 * (1 - x1)
        x2 = 3.9 * (x2 - x2 * x2)
    print("For", iterations, "iternations, x1 is now", x1, "and x2 is now", x2)
main()

This program illustrates a chaotic function
For 23 iternations, x1 is now 0.7988018086971822 and x2 is now 0.7988018086646516
